Dphan/add nats

dorotaphanSiili opened the pull request:
<!-- 
Thanks for creating this pull request 🤗

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple prs instead of opening a huge one.
-->

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes #24

Update the https://github.com/apigear-io/apigear-docs repository with this version (Contains the documentation changes)

dorotaphanSiili committed:
feat(nats): add adapters templates - no nats yet

This commit introduces the basic functionality for nats adapters.
It does not yet uses the nats protocol.
It generates all the methods that will be used by the users of
interface.
The nats functionality will be added with next commits.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): add nats base

This is base class for client and service as they have lot of common
functionality.
This is part of #24 task.

dorotaphanSiili committed:
feat(nats): handle request-response

Prepare the support to use request-response nats mechanism.
It will be necessary for handling methods remote calls.
The functionality added covers subbscribe, unsubscribe, storing callbacks,
Also disconnect added.
The implementation uses similar structures and functions as for subscriptions,
with details different - the callback added for topic has a return value.
In contrary to publish subscribe mechanism which is to be used for properties and signals,
only the requestor gets the response, not all the client adapters for same interface.
The request-response mechanism got its own thread pool in nats base.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): make subscriptions non-blocking

Subscribe was a thread blocking call.
A thread pool was introduced to make those calls in a background.
The subscribe function also gets the subscribe callback which is
executed after the subscription action is finished either successfully
or not.
This will be used by the usesr to track the state of all its subscriptions.

dorotaphanSiili committed:
feat: add publisher helper

Add a templated class as a publisher helper.
Its puropse is to help keep track of subscribers:
it allows subscribing and provide an id to unsubscribe, on publishChange
call it notifies all the subscribers about the change.
It is templated to allow different types of the changed value, and to
publish signals with many parameters.
It does not store any value.
It is thread safe.
The id generating mechanism may be improved.

It will be used in next commit to keep track of the nats adapters state.
It also may be used for generated interface publishers.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): add base adapter

This commit introduces a base class for the generated adapters,
both client and service side.
The functionality it covers:
handling is ready for an adapter- where is ready means that the adapter
has working connection and has sucessfully subscribed to all topics it needed.
It also keeps track of the subscription state.

This is part of th #24 task.

dorotaphanSiili committed:
feat(nats): client adapter implementation

Client Interface adapters now are using nats protocol.
They us base adatapter as a base class.
As the base requires weak_ptr it inherites enabl_shared_from_this.
The weak_ptr is necessary for making sure that callbacks passed to nats
client are always safe to call especially during tear down.
This is also why it has a create method instead of public constructor,
and it has an init method - to which all the logic for initialization
was moved to ensure no virtual method will be called on construction.

For the request response mechanism - if there is no responder the
returned value in an empty string, which causes crash on trying to
parse. This could be solved in a different way - detecting if there is a
service side running, waiting for init perhaps - but it won't work in
all cases as for now like when there was a service but it got
disconnected. If such functionality is required it need to be
implemented covering all connect-disconnect scenarios.

This is part of task #24.

dorotaphanSiili committed:
feat(nats): nats service adapters

Service Interface adapters now are using nats protocol.
They us base adatapter as a base class.
As the base requires weak_ptr it inherites enabl_shared_from_this.
The weak_ptr is necessary for making sure that callbacks passed to nats
client are always safe to call especially during tear down.
This is also why it has a create method instead of public constructor,
and it has an init method - to which all the logic for initialization
was moved to ensure no virtual method will be called on construction.

dorotaphanSiili committed:
feat(nats): add service and client

Add service and client adapters which will be used by generated
intrefaces. This is the interface for them to use functionalities
provided by CNATS.
This commit only adds the skeleton classes. The implementation will be
added in later commits
This is part of #24 task.

dorotaphanSiili committed:
feat(nats): add thread safety to base adapter

intorduces mutexes for stored subscriptions,
ensures that the callbacks won't be used after the object deleted
fixes checking for is ready:
in case the subscriptions are added one by one after the connection was
established the is connected callback was called after each subscription
added (the vector size didn't had its full size yet). Now it is ensured
that the isReady will actually wait for all the subscriptions.

dorotaphanSiili committed:
feat(nats): support debug mode

The CNATS library has a d suffix for debug version.
As far as the cmake easy supports this with linking debug or optimised version, the conan has some issues with it.
For conan build there is the name of the library set in conan file, and later passed to cmake.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): modify build apigear

This commit ensures conan build has a folder structure like for cmake build.
This is necessary for apigear submodules that are using other, sibiling apigear modules.

dorotaphanSiili committed:
feat(nats): add examples for nats

Add client and service side application.
Those are adapters, still an external NATS server needs to be used.
Use it as a starting point, add other calls, emit signals.

This is part of the #24 task.

dorotaphanSiili committed:
feat(nats): handle unsubscribe

Inform the subscriber about the subscription closed.

dorotaphanSiili committed:
feat(nats): expose client Id

this is preparatory change for handiling cilent initialization with service side state
each client obtains the id on connection(valid through that connection
session).
This id will be used for initialization - the init message from service
is send to each client on demand (init request), and the id is in the
message topic.

This is part of the #24 task.

dorotaphanSiili committed:
feat(nats): init nats client with service state

each client subscribes for init message,
after successfull subscription of all topics it sends init request with its id in
payload, the topic contains the module and interface name.
If there is a service, already subscribed for such init request it
serves the message:
sends back one message wiht all the properties, only for this client
(the topic contains the id, and for such topic the client is
subscribed).
For the case when no service at time the client publishes the init request,
there is a messege with the interface service adapter availablity published
by the service once it has subscribed for all topics.
All the client interfaces adapters subscribe for this message.
This does not affect the is_ready functionality - which, as it was only
says the client is connected to NATS server with all its topic
subscribed, and not if it has connection with a service side of
interface.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): handle subscriptions and messages

Adds handling subscriptions, sending and receiving messages.
The subscribe function has a callback for a topic to be handled.
Each subscription - even for the same topic, gets different id.
The ids are used to unsubscribe a callback from a topic.
The subscriptions and message handlers are stored in non-relocable containers.
This is very important, as the pointers to those are passed to NATS handlers.
Also a handler for errors for subscriptions is added during connection configuration.

The callbacks are stored before a subscription is made,
because a context with it is passed in form of a pointer to storage, to the subscribe function.

The subscriptions are wrapped with smart pointer (if successfully allocated).

Each subscription has also set a callback to clean it when it stops being used:
it is guaranteed that no message will be received on this subscription and it is ready to be cleaned.
This happens either after unsubscribe call or when connection gets closed.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): add nats with empty impl

This is a first commit for implementing the nats protocol.

The CNATS libray is used for handling protocol.
The idea is to separate this library, from the code exposed to user, so it will be used only through the apigear/nats wrapper.
This commit introduces almost empty initialization. The functionality will be added with further commits.
The dependency for the CNATS library is added to conan feature.
It is possible to build it with CMake too, but you need to provide the library (you can use conan to get the dependencies only, see the CI workfolw for CMake build)

The requirement for the conan is added with visibility set to false - we wan't it to be build as static library and as a implementation detail, therefore it is not passed any further to user of the apigear/nats users.

This is part of #24 task.

dorotaphanSiili committed:
feat(nats): handle connect

CWrapper now connects to a NATS server to given addres.
The connect function also allows adding a callback to be informed about connection status changes.

All the resources provided by th elibrary are wrapped with smart pointers with custom deleters, as the nats library is one that stores the requested items (connection, options, messages).
The class also stores a connection context, which is passed the CWrapper callbacks. The context contains the self weak_ptr, hence the

An option that is worth mentioning is natsOptions_UseGlobalMessageDelivery, without it set to true, the CNATS library would use separate thread for each subscription (topic). We assume to have a lot of subscriptions, that is why we decided to set it to true, which means, there is some thread pool which will be used instead.

Basic subscription handling was added just for test purposes. Proper implementation will be delivered in coming commits.
This is part of #24 task.

dorotaphanSiili committed:
feat(nats): handle connection in natsbase

This is an interface for using cwrapper.
Handles connection and minds the connection state.
Allows adding callback for connection state changed.

dorotaphanSiili committed:
feat(nats): generated tests

uses new filter cppTestValue which produces non default vale
with this, for each interface there are test to check setting a
property and receiving a signal.
Also all methods are tested with sync and async version - but with
default values. The return value is also default

dorotaphanSiili committed:
feat(nats): add flush

give better control over to user for sending messages.
The connection is configured in a wayt the messages are stored
for a period of time or untill buffer full.
With a flush function exposed, the user can force nats client to send
the messages. Nats offers also option to send right away each message
but it needs to be configured on connection start, and is same for all
the messages see the natsOptions_SetSendAsap. If that is desired
behavior then it should be used instead of flush which is less efficient
as it requires a round-trip with the server.

This is part of #24 task.

dorotaphanSiili committed:
feat(Nats): enable nats test for linux, run nats server

dorotaphanSiili committed:
ci(nats): add nats features

Prepare CI to run nats tests on linux os (easy setup for server).
Install CNats with conan works well, but installing it with cmake
requires passing extra header- the library with all extra feature off
(like streaming) fails to build.

dorotaphanSiili committed:
feat(nats): configure connection

Adds option to set sending each message as soon as possibe without
waiting for the buffer to be filled (or max time to pass).
As default the option is set to true as for local tests its latency and
test time even for lot of messages are way smaller.
Depending on your requirements for application you may want to set this
option to false.

dorotaphanSiili committed:
feat(nats): add documentation

dorotaphanSiili committed:
feat(docs): add class description

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+:::
+
+:::note
+A [CNATS](https://nats-io.github.io/nats.c/) library is used to implement NATS in this library.
```
minor wording: The CNATS library is used....

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+A [CNATS](https://nats-io.github.io/nats.c/) library is used to implement NATS in this library.
+:::
+
+This feature provides NATS protocol adapters into your project, and also show that an existing protocol can be adapted for sharing your data in your ecosystem. When going through this document you may notice this implementation contains general client/server adapters in `📂hello-world/apigear/nats`
```
what do these adapters do? Also I think the wording is: provides ... for your project.

dorotaphanSiili replied:
Changed: This feature provides NATS protocol adapters into your project, and also show that an existing
Inot: This feature provides NATS protocol adapters for your objects defined by interfaces. With this feature you can also see how
I hope it is more clear now :)

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+A [CNATS](https://nats-io.github.io/nats.c/) library is used to implement NATS in this library.
+:::
+
+This feature provides NATS protocol adapters into your project, and also show that an existing protocol can be adapted for sharing your data in your ecosystem. When going through this document you may notice this implementation contains general client/server adapters in `📂hello-world/apigear/nats`
```
"and also show that an existing protocol can be adapted for sharing your data in your ecosystem." Not sure what this means. protocol - adapted - data - ecosystem.

dorotaphanSiili replied:
It is- like with mqtt - to show that users can wite such feature by themself to adapt any existing protocol like whatever they use in their ecosystem, like it doesn't have to be olink which is not production proven. So this is example of how it can be done

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+- Use an _Nats server adapter_ to expose your interface implementation as a remote service. You can expose same service for different technologies (olink, mqtt, other adapted protocol).
+
+:::tip
+The NATS server is not provided with the template. To be able to run apigear client and service (both serve as NATS clients) you need to run a server of your choice, which is accessible for both client and service.
```
There is only one nats server (nats.io)

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+
+:::caution
+In this library we use STL classes in public interfaces.
+We assume that this library is built from source (configuration provided by generated `CMakeLists`).
```
We assume that this library is built from source using cmake

dorotaphanSiili replied:
I'd rather keep it in a form it is in, we use also conan- true it does use cmake but I am not sure about the wording 🤔

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+			it allows requesting call (with returning a message), subscribing for topics and publishing them. It also allows subscription and conntection state monitoring.
+:::note
+The CNATS allows connection configuration, please check [nats options](https://nats-io.github.io/nats.c/group__opts_group.html) for more information.
+The most important options used in this implementation, that may influence the performance are SetSendAsap and UseGlobalMessageDelivery, both set to true.
```
SetSendAsap and UseGlobalMessageDelivery should be in code blocks (e.g. ``)

jryannel commented on docs/docs/features/nats.md:
```diff
@@ -0,0 +1,231 @@
+
+Files `📜helloclient.h` and `📜helloclient.cpp` implement the remote client for the `Hello` interface - a `HelloClient` class.<br />
+The object is an `IHello` implementation.<br />
+It requires an instance of ApiGear::Nats::Client to work. It uses the Client to subscribe (and unsubscribe) for topics that allow receiving properties, signals and invoke responses from service.
```
code block

jryannel commented on goldenmaster/modules/tb_enum/generated/nats/enuminterfaceservice.h:
```diff
@@ -0,0 +1,60 @@
+    /// @brief requests to set the value for the property Prop3 coming from the client
+    /// @param fields contains the param of the type Enum3Enum
+    void onSetProp3(const std::string& args) const;
+std::string onInvokeFunc0(const std::string& args) const;
```
spacing

jryannel reviewed (approved):
some minor wording and spacing. Looks great. Nice work.

dorotaphanSiili committed:
fix(nats): change utilities lib availability

apigear nats library uses apigear utilities, its visibility was changed
to public, so its users also can use it.
Since it is standard and common part of apigear is should be available
also for the users.
